src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/EvidenceTargetLinkClusterer.java

```diff
@@ -0,0 +1,122 @@
+import org.broadinstitute.hellbender.tools.spark.sv.utils.SVInterval;
+import org.broadinstitute.hellbender.tools.spark.sv.utils.SVIntervalTree;
+import org.broadinstitute.hellbender.utils.Utils;
+import scala.Tuple2;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class EvidenceTargetLinkClusterer {
+
+    private final ReadMetadata readMetadata;
+
+    public EvidenceTargetLinkClusterer(final ReadMetadata readMetadata) {
+        this.readMetadata = readMetadata;
+    }
+
+    public Iterator<EvidenceTargetLink> cluster(final Iterator<BreakpointEvidence> breakpointEvidenceIterator) throws Exception {
+        final List<EvidenceTargetLink> links = new ArrayList<>();
+        final SVIntervalTree<EvidenceTargetLink> currentIntervalsWithTargets = new SVIntervalTree<>();
+        while (breakpointEvidenceIterator.hasNext()) {
+            final BreakpointEvidence nextEvidence = breakpointEvidenceIterator.next();
+            if (nextEvidence.hasDistalTargets(readMetadata)) {
+                final SVInterval target = nextEvidence.getDistalTargets(readMetadata).get(0);
+                EvidenceTargetLink updatedLink = null;
+                for (final Iterator<SVIntervalTree.Entry<EvidenceTargetLink>> it = currentIntervalsWithTargets.overlappers(nextEvidence.getLocation()); it.hasNext(); ) {
+                    final SVIntervalTree.Entry<EvidenceTargetLink> sourceIntervalEntry = it.next();
+                    final EvidenceTargetLink oldLink = sourceIntervalEntry.getValue();
+                    // todo: what to do if there are more than one distal targets -- for now just taking the first one
+                    if (nextEvidence.hasDistalTargets(readMetadata) &&
+                            strandsMatch(nextEvidence.isForwardStrand(), sourceIntervalEntry.getValue().sourceForwardStrand)
+                        && (nextEvidence.getDistalTargets(readMetadata).get(0).overlaps(oldLink.target) &&
+                                strandsMatch(nextEvidence.getDistalTargetStrands(readMetadata).get(0), oldLink.targetForwardStrand))) {
+                            // if it does, intersect the source and target intervals to refine the link
+                            it.remove();
+                            final SVInterval newSource = sourceIntervalEntry.getInterval().intersect(nextEvidence.getLocation());
+                            final SVInterval newTarget = oldLink.target.intersect(nextEvidence.getDistalTargets(readMetadata).get(0));
+                            updatedLink = new EvidenceTargetLink(newSource,
+                                    oldLink.sourceForwardStrand,
+                                    newTarget,
+                                    oldLink.targetForwardStrand,
+                                    nextEvidence instanceof BreakpointEvidence.DiscordantReadPairEvidence
```

Comment:
I don't think so.. These ternaries are hard to read but in this clause and the next I'm saying:

- if it's a read pair use the old split read count, otherwise increment the split read count
- if it's a read pair increment the read pair count, otherwise use the old read pair count

Maybe if I pull these out into named variables it will be clearer.